### PR TITLE
Add Variational Autoencoder and GANs

### PR DESCRIPTION
**Pull request by sarthak247**
Add Variational Autoencoder and GANs part for Unit 5 - Generative Models w.r.t #28 

Review required. :v:

### CODE REVIEW

File: chapters/en/Unit 5 - Generative Models/gans.mdx [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Generative Adversarial Networks
+
+## Section 2.1 Introduction to GANs
+Generative Adversarial Networks (GANs) are a class of deep learning models introduced by [Ian Goodfellow](https://scholar.google.ca/citations?user=iYN86KEAAAAJ&hl=en) and his colleagues in 2014. The core idea behind GANs is to train a generator network to produce data that is indistinguishable from real data, while simultaneously training a discriminator network to differentiate between real and generated data.
```
Comment (ariG23498) [inline chapters/en/Unit 5 - Generative Models/gans.mdx:4]: Sometimes the `discriminator` is refered to as the `critic`. I have always liked this term better as this directly comes from the analogy of a painter (generator) and an art critic (discriminator). This analogy was part of the [GAN course from DeepLearning.ai](https://www.deeplearning.ai/courses/generative-adversarial-networks-gans-specialization/).

This is not a suggestion. Might be interesting to pursue this analogy in the introductory paragraph. Really drives the intuition home.

File: chapters/en/Unit 5 - Generative Models/gans.mdx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    - **VAEs:**
+        * **Strengths:** Easier to train and more stable than GANs
+        * **Weaknesses:** May generate blurry, less detailed images with unrealistic features.
+* **Other Tasks:**
+    - **GANs:**
+        * **Strengths:** Can be used for tasks like super-resolution and image-to-image translation.
+        * **Weaknesses:** May not be the best choice for tasks that require a smooth transition between data points
+    - **VAEs:**
+        * **Strengths:** Widely used for tasks like image denoising and anamoly detection.
+        * **Weaknesses:** May not be as effective as GANs for tasks that require high-quality image generation.
```
Comment (ariG23498) [inline chapters/en/Unit 5 - Generative Models/gans.mdx:28]: I would always cite different papers to prove the pointers mentioned. If not papers, reports of experiments conducted.

File: chapters/en/Unit 5 - Generative Models/gans.mdx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+Here's a table summarizing the key differences:
+|Feature|GANs|VAEs|
+|-------|-----|---|
+|Image Quality|Higher|Lower|
+|Ease of Training|More difficult|Easier|
+|Stability|Less Stable|More Stable|
+|Applications|Image Generation, Super-resolution, image-to-image translation|Image Denoising, Anamoly Detection, Signal Analysis|
```
Comment (ariG23498) [inline chapters/en/Unit 5 - Generative Models/gans.mdx:36]: This is so neat! A summarizer of all the pointers is a very thoughtful.

File: chapters/en/Unit 5 - Generative Models/gans.mdx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+Ultimately, the best choice depends on one's specific needs and priorities. If one need high-quality images for tasks like generating realistic faces or landscapes, then a GAN might be the better choice. However, if one needs a model that is easier to train and more stable, then a VAE might be a better option.
+
+## Section 2.3 Training GANs
```
Comment (ariG23498) [inline chapters/en/Unit 5 - Generative Models/gans.mdx:40]: [Slide 12: Using Math to Communicate](https://docs.google.com/presentation/d/e/2PACX-1vRau7IXhvdKKOZQt3hujyw5dNltExsmxfHTMbHTpwUXK4mst-lXMPkLQL8MRNsVVONWRgBdTf5W3k4c/pub?start=false&loop=false&delayms=3000&slide=id.g157b6906e85_0_25)
[Slide 13: The Samplings](https://docs.google.com/presentation/d/e/2PACX-1vRau7IXhvdKKOZQt3hujyw5dNltExsmxfHTMbHTpwUXK4mst-lXMPkLQL8MRNsVVONWRgBdTf5W3k4c/pub?start=false&loop=false&delayms=3000&slide=id.g157b6906e85_0_38)
[Slide 14: Discriminator](https://docs.google.com/presentation/d/e/2PACX-1vRau7IXhvdKKOZQt3hujyw5dNltExsmxfHTMbHTpwUXK4mst-lXMPkLQL8MRNsVVONWRgBdTf5W3k4c/pub?start=false&loop=false&delayms=3000&slide=id.g157b6906e85_0_50)
[Slide 15: Generator](https://docs.google.com/presentation/d/e/2PACX-1vRau7IXhvdKKOZQt3hujyw5dNltExsmxfHTMbHTpwUXK4mst-lXMPkLQL8MRNsVVONWRgBdTf5W3k4c/pub?start=false&loop=false&delayms=3000&slide=id.g157bb6d6115_0_2)
[Slide 16: Revisited](https://docs.google.com/presentation/d/e/2PACX-1vRau7IXhvdKKOZQt3hujyw5dNltExsmxfHTMbHTpwUXK4mst-lXMPkLQL8MRNsVVONWRgBdTf5W3k4c/pub?start=false&loop=false&delayms=3000&slide=id.g157bb6d6115_0_19)

These might help you build a better intuition about the min max game.

A [desmos](https://www.desmos.com/calculator/nw6m38xtij) demo for the mathametical equation also helps here.

**Review by ariG23498 — CHANGES_REQUESTED** [review]
This is a very thorough write up for GANs. I am yet to review the VAEs.

**Review by merveenoyan — COMMENTED** [review]
I only left formatting nits, thank you!

File: chapters/en/Unit 5 - Generative Models/gans.mdx (hunk 2 above)
Comment (sarthak247) [reply chapters/en/Unit 5 - Generative Models/gans.mdx:28]: I got these points after some Google search and from some medium posts. I don't know of any papers which directly reference a comparison like this between GANs and VAEs so if you know of any paper do let me know and I will add it. Cheers :v:

Comment (sarthak247) [reply chapters/en/Unit 5 - Generative Models/gans.mdx:28]: I think this should work. Although I'm not sure about the sources
* [What are the fundamental differences between VAE and GAN for image generation?](https://ai.stackexchange.com/questions/25601/what-are-the-fundamental-differences-between-vae-and-gan-for-image-generation)
* [Issues with GAN and VAE models](https://stats.stackexchange.com/questions/541775/issues-with-gan-and-vae-models)
* [VAE Vs. GAN For Image Generation](https://www.baeldung.com/cs/vae-vs-gan-image-generation)
* [Diffusion Models vs. GANs vs. VAEs: Comparison of Deep Generative Models](https://towardsai.net/p/machine-learning/diffusion-models-vs-gans-vs-vaes-comparison-of-deep-generative-models)

File: chapters/en/Unit 5 - Generative Models/gans.mdx (hunk 1 above)
Comment (sarthak247) [reply chapters/en/Unit 5 - Generative Models/gans.mdx:4]: Added

File: chapters/en/Unit 5 - Generative Models/gans.mdx [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# Generative Adversarial Networks
+
+## Section 2.1 Introduction to GANs
```
[the remark this answers is not in the document]
Comment (sarthak247) [reply chapters/en/Unit 5 - Generative Models/gans.mdx]: I've resolved these grammatical errors. However, if there are any remaining please do let me know. Cheers :v:

File: chapters/en/Unit 5 - Generative Models/gans.mdx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+* **Architecture overview:** GANs consist of two main components: `the generator` and `the discriminator`
+* **Generator:** The generator takes random noise $$z$$ as input and generates synthetic data samples. Its goal is to create data that is realistic enough to deceive the discriminator.
+* **Discriminator:** The discriminator, akin to a detective, evaluates whether a given sample is real (from the actual dataset) or fake (generated by the generator). Its objective is to become increasingly accurate in distinguishing between real and generated samples.
+
+A common analogy which can be found online is that of an art forger/painter (the geneator) which tries to forge paintings and an art investigator/critic (the discriminator) which triest to detect imitations
+
+![Lilian Weng GAN Figure](https://lilianweng.github.io/posts/2017-08-20-gan/GAN.png)
+
+## Section 2.2 GANs vs VAEs
```
Comment (merveenoyan) [inline chapters/en/Unit 5 - Generative Models/gans.mdx:13]:
```suggestion
## GANs vs VAEs
```
No need for numbering

File: chapters/en/Unit 5 - Generative Models/variational_autoencoders.mdx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+![Vanilla Autoencoder Image - Lilian Weng Blog](https://lilianweng.github.io/posts/2018-08-12-vae/autoencoder-architecture.png)
+
+This encoder model consists of an encoder network (represented as $$\g_\phi$$) and a decoder network (represented as $$f_\theta$$).
+The low-dimensional representation is learned in the bottleneck layer as $$z$$ and the reconstructed output is represented as $$x'=f_\theta(g_\phi(x))$$ with the goal as $$x\approx x'$$. 
+A common loss function used in such vanilla autoencoders is $$L(\theta, \phi) = \frac{1}{n}\sum_{i=1}^n (\mathbf{x}^{(i)} - f_\theta(g_\phi(\mathbf{x}^{(i)})))^2$$ with tries to minimize the error between the original image and the reconstructed one and is also known as the `reconstruction loss`
+Autoencoders are useful for tasks such as data denoising, feature learning, and compression. However, traditional autoencoders lack the probabilistic nature that makes VAEs particularly intriguing and also useful for generational tasks
+
+## Variational Autoencoders (VAEs) Overview:
```
Comment (merveenoyan) [inline chapters/en/Unit 5 - Generative Models/variational_autoencoders.mdx:15]:
```suggestion
## Variational Autoencoders (VAEs) Overview
```

File: chapters/en/Unit 5 - Generative Models/variational_autoencoders.mdx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+![Autoencoders vs VAEs - Sciforce Medium](https://miro.medium.com/v2/resize:fit:786/format:webp/0*FODurdX4jiWz5bKS)
+In the context of Vanilla Autoencoders (AE), the smile feature is encapsulated as a fixed, deterministic value. In contrast, Variational Autoencoders (VAEs) are deliberately crafted to encapsulate this feature as a probabilistic distribution. This design choice facilitates the introduction of variability in generated images by enabling the sampling of values from the specified probability distribution.
+
+## Section 1.3: Mathematics Behind VAEs
+Understanding the mathematical concepts behind VAEs involves grasping the principles of probabilistic modeling and variational inference.
+![Variational Autoencoder - Lilian Weng Blog](https://lilianweng.github.io/posts/2018-08-12-vae/vae-gaussian.png)
```
Comment (merveenoyan) [inline chapters/en/Unit 5 - Generative Models/variational_autoencoders.mdx:26]: You can add the images to hf-vision/course-assets and give a reference so even if the image is reviewed from the source it won't be removed from here

File: chapters/en/Unit 5 - Generative Models/gans.mdx [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        ![Example of GAN-Generated bedrooms taken from Unsupervised Representation Learning with Deep Convolutional Generative Adversarial Networks, 2015](https://machinelearningmastery.com/wp-content/uploads/2019/04/Example-of-GAN-Generated-Photographs-of-Bedrooms.png)
+    - **VAEs:**
+        * **Strengths:** Easier to train and more stable than GANs
+        * **Weaknesses:** May generate blurry, less detailed images with unrealistic features.
+* **Other Tasks:**
+    - **GANs:**
+        * **Strengths:** Can be used for tasks like super-resolution and image-to-image translation.
+        * **Weaknesses:** May not be the best choice for tasks that require a smooth transition between data points
+    - **VAEs:**
+        * **Strengths:** Widely used for tasks like image denoising and anamoly detection.
```
Comment (merveenoyan) [inline chapters/en/Unit 5 - Generative Models/gans.mdx:29]:
```suggestion
        * **Strengths:** Widely used for tasks like image denoising and anomaly detection.
```

File: chapters/en/Unit 5 - Generative Models/gans.mdx [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+Here's a table summarizing the key differences:
+|Feature|GANs|VAEs|
+|-------|-----|---|
+|Image Quality|Higher|Lower|
+|Ease of Training|More difficult|Easier|
+|Stability|Less Stable|More Stable|
+|Applications|Image Generation, Super-resolution, image-to-image translation|Image Denoising, Anamoly Detection, Signal Analysis|
+
+Ultimately, the best choice depends on one's specific needs and priorities. If one need high-quality images for tasks like generating realistic faces or landscapes, then a GAN might be the better choice. However, if one needs a model that is easier to train and more stable, then a VAE might be a better option.
```
Comment (merveenoyan) [inline chapters/en/Unit 5 - Generative Models/gans.mdx:40]:
```suggestion
Ultimately, the best choice depends on one's specific needs and priorities. If one needs high-quality images for tasks like generating realistic faces or landscapes, then a GAN might be the better choice. However, if one needs a model that is easier to train and more stable, then a VAE might be a better option.
```

File: chapters/en/Unit 5 - Generative Models/gans.mdx [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+Generative Adversarial Networks (GANs) are a class of deep learning models introduced by [Ian Goodfellow](https://scholar.google.ca/citations?user=iYN86KEAAAAJ&hl=en) and his colleagues in 2014. The core idea behind GANs is to train a generator network to produce data that is indistinguishable from real data, while simultaneously training a discriminator network to differentiate between real and generated data.
+* **Architecture overview:** GANs consist of two main components: `the generator` and `the discriminator`
+* **Generator:** The generator takes random noise $$z$$ as input and generates synthetic data samples. Its goal is to create data that is realistic enough to deceive the discriminator.
+* **Discriminator:** The discriminator, akin to a detective, evaluates whether a given sample is real (from the actual dataset) or fake (generated by the generator). Its objective is to become increasingly accurate in distinguishing between real and generated samples.
+
+A common analogy which can be found online is that of an art forger/painter (the geneator) which tries to forge paintings and an art investigator/critic (the discriminator) which triest to detect imitations
```
Comment (merveenoyan) [inline chapters/en/Unit 5 - Generative Models/gans.mdx:9]:
```suggestion
A common analogy that can be found online is that of an art forger/painter (the generator) which tries to forge paintings and an art investigator/critic (the discriminator) who tries to detect imitations.
```

File: chapters/en/Unit 5 - Generative Models/gans.mdx [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+![Lilian Weng GAN Figure](https://lilianweng.github.io/posts/2017-08-20-gan/GAN.png)
+
+## Section 2.2 GANs vs VAEs
+GANs and VAEs are both popular generative models in the field of machine learning, but they have different strengths and weaknesses. Whether one is "better" than the other depends on the specific task and requirements. Here's a breakdown of their strengths and weaknesses.
```
Comment (merveenoyan) [inline chapters/en/Unit 5 - Generative Models/gans.mdx:14]:
```suggestion
GANs and VAEs are both popular generative models in machine learning, but they have different strengths and weaknesses. Whether one is "better" depends on the specific task and requirements. Here's a breakdown of their strengths and weaknesses.
```

**Review by merveenoyan — COMMENTED** [review]
This is very close to merger! Just left formatting and typo corrections 🤗 
also @ariG23498 can you leave another review? @sarthak247 has iterated on the content and we'll have to merge next

**Review by ariG23498 — APPROVED** [review]
Looks good to me!

Love how crisp and to the point this is. 🤗

File: chapters/en/Unit 5 - Generative Models/gans.mdx [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+* **Image Generation:**
+    - **GANs:**
+        * **Strengths:** Generate higher quality images, especially for complex data with sharp details and realistic textures.
```
[the remark this answers is not in the document]
Comment (sarthak247) [reply chapters/en/Unit 5 - Generative Models/gans.mdx:17]: I've addressed the other issues but I believe removing the indentations might impact the structure of the content. Please let me know whether it should still be removed or kept.

File: chapters/en/Unit 5 - Generative Models/variational_autoencoders.mdx (hunk 8 above)
Comment (sarthak247) [reply chapters/en/Unit 5 - Generative Models/variational_autoencoders.mdx:26]: I'll update them as soon as it gets merged to hf-vision/course-assets

**Comment by sarthak247** [comment]
I've made the requested changes :v:

**Comment by merveenoyan** [comment]
@sarthak247 can you check out rest of the chapters and if it doesn't exist, maybe add a very minimal PyTorch implementation of a GAN and a VAE?

**Review by merveenoyan — APPROVED** [review]
@sarthak247 I'll merge this one, you can raise another PR for the implementation if it's o 🤗 thanks a lot!

**Pull request merged**